File: src/main/go/io/github/seeflood/copy-paste-ds/tree/segment/SegmentTree.go
package segment_tree

import "errors"

type Tree interface {
	Query(lo int, hi int) (interface{}, error)
	Set(lo int, hi int, value interface{}) error
	Update(lo int, hi int, value interface{}) error
}
type AggregateFunction interface {
	aggregate(a interface{}, b interface{}) interface{}
	update(old interface{}, from int, to int, diff interface{}) interface{}
	set(from int, to int, newValue interface{}) interface{}
	getEmpty() interface{}
}

type Adder struct {
}

func (adder *Adder) aggregate(a interface{}, b interface{}) interface{} {
	return a.(int) + b.(int)
}

func (adder *Adder) update(old interface{}, from int, to int, diff interface{}) interface{} {
	return old.(int) + diff.(int)*(to-from+1)
}

func (adder *Adder) set(from int, to int, newValue interface{}) interface{} {
	return newValue.(int) * (to - from + 1)
}

func (adder *Adder) getEmpty() interface{} {
	return 0
}

func CreateSegmentTree(size int, aggregator AggregateFunction) Tree {
	// 1. init tree array
	arrSize := 2*nextPowerOf2(size) - 1
	if arrSize < 0 {
		arrSize = 0
	}
	arr := make([]interface{}, arrSize)
	for i := 0; i < arrSize; i++ {
		arr[i] = aggregator.getEmpty()
	}
	lazy := make([]interface{}, arrSize)
	lazyIsUpdate := make([]bool, arrSize)

	return &segmentTreeImpl{aggregator, arr, lazy, lazyIsUpdate, size}
}

func nextPowerOf2(i int) int {
	if i&(i-1) == 0 {
		return i
	}
	for i&(i-1) != 0 {
		i &= (i - 1)
	}
	return i << 1
}

type segmentTreeImpl struct {
	aggregator   AggregateFunction
	tree         []interface{}
	lazy         []interface{}
	lazyIsUpdate []bool
	n            int
}

func (st *segmentTreeImpl) Query(lo int, hi int) (interface{}, error) {
	return st.query(lo, hi, 0, st.n-1, 0)
}

func (st *segmentTreeImpl) query(qlo int, qhi int, lo int, hi int, idx int) (interface{}, error) {
	if qlo > qhi {
		return nil, errors.New("Illegal params.")
	}
	if hi < qlo || qhi < lo {
		return st.aggregator.getEmpty(), nil
	}
	if qlo <= lo && hi <= qhi {
		return st.nil2Empty(st.tree[idx]), nil
	}
	mid := lo + (hi-lo)/2
	left, _ := st.query(qlo, qhi, lo, mid, 2*idx+1)
	right, _ := st.query(qlo, qhi, mid+1, hi, 2*idx+2)
	return st.aggregator.aggregate(left, right), nil
}

func (st *segmentTreeImpl) nil2Empty(v interface{}) interface{} {
	if v == nil {
		return st.aggregator.getEmpty()
	}
	return v
}

func (st *segmentTreeImpl) Set(lo int, hi int, value interface{}) error {
	return st.update(lo, hi, 0, st.n-1, 0, value, false)
}

func (st *segmentTreeImpl) Update(lo int, hi int, value interface{}) error {
	return st.update(lo, hi, 0, st.n-1, 0, value, true)
}

func (st *segmentTreeImpl) update(qlo int, qhi int, lo int, hi int, idx int, value interface{}, isUpdate bool) error {
	if qlo > qhi {
		return errors.New("Illegal params.")
	}
	if lo > hi {
		return nil
	}
	if hi < qlo || qhi < lo {
		return nil
	}
	st.checkLazy(lo, hi, idx)
	//	cover
	if qlo <= lo && hi <= qhi {
		if isUpdate {
			st.lazyIsUpdate[idx] = true
		} else {
			st.lazyIsUpdate[idx] = false
		}
		st.lazy[idx] = value
		st.checkLazy(lo, hi, idx)
		return nil
	}
	//	overlap
	mid := lo + (hi-lo)/2
	leftI := 2*idx + 1
	rightI := 2*idx + 2
	st.update(qlo, qhi, lo, mid, leftI, value, isUpdate)
	st.update(qlo, qhi, mid+1, hi, rightI, value, isUpdate)
	st.tree[idx] = st.aggregator.aggregate(st.tree[leftI], st.tree[rightI])
	return nil
}

func (st *segmentTreeImpl) checkLazy(lo int, hi int, idx int) {
	if st.lazy[idx] == nil {
		return
	}
	// apply
	st.applyLazy(lo, hi, idx)

	// propagate and merge
	if lo < hi {
		leftI := 2*idx + 1
		rightI := 2*idx + 2
		if st.lazyIsUpdate[idx] {
			st.lazy[leftI] = st.aggregator.aggregate(st.lazy[leftI], st.lazy[idx])
			st.lazy[rightI] = st.aggregator.aggregate(st.lazy[rightI], st.lazy[idx])
		} else {
			st.lazy[leftI] = st.lazy[idx]
			st.lazyIsUpdate[leftI] = false
			st.lazy[rightI] = st.lazy[idx]
			st.lazyIsUpdate[rightI] = false
		}
	}
	//	clear
	st.lazy[idx] = nil
}

func (st *segmentTreeImpl) applyLazy(lo int, hi int, idx int) {
	if st.lazy[idx] == nil {
		return
	}
	if st.lazyIsUpdate[idx] {
		st.tree[idx] = st.aggregator.update(st.tree[idx], lo, hi, st.lazy[idx])
	} else {
		st.tree[idx] = st.aggregator.set(lo, hi, st.lazy[idx])
	}
}

File: java/src/main/java/io/github/seeflood/advanced/ds/SegmentTree.java
package io.github.seeflood.advanced.ds;

/**
 * 线段树
 *
 * @param <T>
 */
public class SegmentTree<T> {
    /**
     * 内置的聚合函数
     *
     * @param <T>
     */
    public static class Functions<T> {
        public static AggregateFunction<Integer> MIN_INTEGER = new AggregateFunction<Integer>() {
            @Override
            public Integer aggregate(Integer a, Integer b) {
                return Math.min(a, b);
            }

            @Override
            public Integer update(Integer old, Integer diff) {
                return old + diff;
            }

            @Override
            public Integer getEmpty() {
                return 0;
            }
        };
        public static AggregateFunction<Integer> MAX_INTEGER = new AggregateFunction<Integer>() {
            @Override
            public Integer aggregate(Integer a, Integer b) {
                if (a == null) {
                    return b;
                }
                if (b == null) {
                    return a;
                }
                return Math.max(a, b);
            }

            @Override
            public Integer update(Integer old, Integer diff) {
                if (old == null) {
                    return diff;
                }
                if (diff == null) {
                    return old;
                }
                return old + diff;
            }

            @Override
            public Integer getEmpty() {
                return 0;
            }
        };
        public static AggregateFunction<Integer> SUM_INTEGER = new AggregateFunction<Integer>() {

            @Override
            public Integer aggregate(Integer a, Integer b) {
                if (a == null && b == null) {
                    return 0;
                }
                if (a == null) {
                    return b;
                }
                if (b == null) {
                    return a;
                }
                return a + b;
            }

            @Override
            public Integer update(Integer old, Integer diff) {
                if (old == null && diff == null) {
                    return 0;
                }
                if (old == null) {
                    return diff;
                }
                if (diff == null) {
                    return old;
                }
                return old + diff;
            }

            @Override
            public Integer getEmpty() {
                return 0;
            }
        };
    }


    /**
     * 聚合函数接口
     *
     * @param <T>
     */
    public interface AggregateFunction<T> {
        T aggregate(T a, T b);

        T update(T old, T diff);

        T getEmpty();
    }

    private final AggregateFunction<T> func;
    private final Object[] st;
    private final Object[] lazy;
    private final boolean[] lazyIsUpdate;
    private final int n;

    public SegmentTree(T[] arr, AggregateFunction<T> func) {
        this.func = func;
        this.n = arr.length;
        int size = 2 * nextPowerOf2(n) - 1;
        this.st = new Object[size];
        this.lazy = new Object[size];
        this.lazyIsUpdate = new boolean[size];
        constructTree(arr, 0, n - 1, 0);
    }

    public T query(int from, int to) {
        return query(from, to, 0, n - 1, 0);
    }

    private T query(int qlo, int qhi, int lo, int hi, int idx) {
        //        check arguments
        if (qlo > qhi) {
            throw new IllegalArgumentException("from can't be larger than to");
        }
//        1. cover
        if (qlo <= lo && hi <= qhi) {
            return castIfNotNull(st[idx]);
        }
//        2. non-overlap
        if (qhi < lo || hi < qlo) {
            return func.getEmpty();
        }
//        3. overlap
        int mid = getMid(lo, hi);
        return func.aggregate(query(qlo, qhi, lo, mid, getLeftChild(idx)), query(qlo, qhi, mid + 1, hi, getRightChild(idx)));
    }

    public void set(int index, T value) {
        update(index, index, 0, n - 1, 0, value, false);
    }

    public void set(int from, int to, T value) {
        //        check arguments
        if (from > to) {
            throw new IllegalArgumentException("from can't be larger than to");
        }
        update(from, to, 0, n - 1, 0, value, false);
    }

    private void update(int qlo, int qhi, int lo, int hi, int idx, T value, boolean isUpdate) {
        if (lo > hi) {
            return;
        }
        propagateLazyToChildren(lo, hi, idx);

//        1. cover
        if (qlo <= lo && hi <= qhi) {
            if (isUpdate) {
                st[idx] = func.update(castIfNotNull(st[idx]), value);
            } else {
//                set value
                st[idx] = value;
            }
//             store lazy flag
            lazy[idx] = value;
            lazyIsUpdate[idx] = isUpdate;
            return;
        }
//        2. non-overlap
        if (qhi < lo || hi < qlo) {
            return;
        }
//        3. overlap
        int mid = getMid(lo, hi);
        int leftChild = getLeftChild(idx);
        update(qlo, qhi, lo, mid, leftChild, value, isUpdate);
        int rightChild = getRightChild(idx);
        update(qlo, qhi, mid + 1, hi, rightChild, value, isUpdate);
        st[idx] = func.aggregate(castIfNotNull(st[leftChild]), castIfNotNull(st[rightChild]));
    }

    private void propagateLazyToChildren(int lo, int hi, int idx) {
        if (lo >= hi || lazy[idx] == null) {
            return;
        }
        int leftChild = getLeftChild(idx);
        int rightChild = getRightChild(idx);
//        apply lazy operation to children
        if (lazyIsUpdate[idx]) {
            st[leftChild] = func.update(castIfNotNull(st[leftChild]), castIfNotNull(lazy[idx]));
            st[rightChild] = func.update(castIfNotNull(st[rightChild]), castIfNotNull(lazy[idx]));
        } else {
            st[leftChild] = lazy[idx];
            st[rightChild] = lazy[idx];
        }
//        set children's lazy flag
        lazy[leftChild] = lazy[idx];
        lazyIsUpdate[leftChild] = lazyIsUpdate[idx];
        lazy[rightChild] = lazy[idx];
        lazyIsUpdate[rightChild] = lazyIsUpdate[idx];
//        remove flag since has been used.
        lazy[idx] = null;
    }

    private int getRightChild(int idx) {
        return 2 * idx + 2;
    }

    private int getLeftChild(int idx) {
        return 2 * idx + 1;
    }

    private void constructTree(T[] arr, int lo, int hi, int idx) {
        if (lo > hi) {
            return;
        }
        if (lo == hi) {
//            set
            st[idx] = arr[lo];
            return;
        }
        int mid = getMid(lo, hi);
        int left = 2 * idx + 1;
        constructTree(arr, lo, mid, left);
        int right = 2 * idx + 2;
        constructTree(arr, mid + 1, hi, right);
        st[idx] = func.aggregate(castIfNotNull(st[left]), castIfNotNull(st[right]));
    }

    private T castIfNotNull(Object o) {
        if (o == null) {
            return null;
        }
        return (T) o;
    }

    private int getMid(int lo, int hi) {
        return lo + (hi - lo) / 2;
    }

    private int nextPowerOf2(int i) {
        if (i <= 0) {
            return 0;
        }
        if ((i & (i - 1)) == 0) {
            return i;
        }
        while ((i & (i - 1)) != 0) {
            i &= (i - 1);
        }
        return i << 1;
    }

}
